Query node predictable ids

ondratra opened the pull request:
This PR adds predictable IDs to entities that previously hadn't predictable IDs. Partially solves #2411 - problem solved for Sumer release, Olympia not covered in this PR.

ondratra committed:
query node - predictable ids

ondratra committed:
query node - predictable ids II

mnaamani commented on query-node/mappings/src/common.ts:
```diff
@@ -40,20 +41,36 @@ export function invalidMetadata(extraInfo: string, data?: unknown): void {
+    ? content
+    : JSON.stringify(content)
+
+  const id = `${event.blockNumber}_${event.index}_${contentType}`
```
If the event index is relative to the transaction and not the block, this is not sufficiently deterministic if there happen to be two extrinsics in the same block that trigger an event at the same index with the same content.
If we go with this id implementation, we should probably add the transaction index as well.

ondratra replied:
Id generation mechanism has been changed and is using simple number that is incremented each time (stored in db as `NextEntityId` entity).

ondratra committed:
Merge branch 'sumer' into query_node_predictable_ids

ondratra committed:
Merge branch 'sumer' into query_node_predictable_ids

ondratra committed:
query node - predictable ids III

mnaamani commented:
I think there is some left over code from previous approach of trying to make deterministic IDs, where you are passing the entire event instead of just just event.blockNumber as argument. It doesn't seem necessary as part of this PR, is it useful to keep?

mnaamani commented on query-node/mappings/src/common.ts:
```diff
@@ -41,20 +42,41 @@ export function invalidMetadata(extraInfo: string, data?: unknown): void {
+/*
+  Creates a predictable and unique ID for the given content.
+*/
+export async function createPredictableId(db: DatabaseManager): Promise<string> {
```
The term predictable id is not the best name, why not simply.

```suggestion
export async function getNextId(db: DatabaseManager): Promise<string> {
```

ondratra replied:
The function has been renamed to `getNextId`.

mnaamani commented on query-node/mappings/src/content/video.ts:
```diff
@@ -465,6 +467,14 @@ function integrateVideoMediaMetadata(
   mediaMetadata.encoding = encoding
 
+  // ensure predictable ids
+  if (!mediaMetadata.encoding.id) {
```
Does this test really work? What if the id is `0`

ondratra replied:
Yes this could cause a trouble in a rare case. It has been prevented by setting mining `getNextId` result to `1`.

ondratra committed:
query node - predictable ids IV

ondratra committed:
query node - predictable ids V - dead code removal

ondratra commented:
Yes, there was a leftover code. I removed it for this PR, but a similar change (reading from `event`) was also introduced by other PRs. So if you see a similar code in other PR reviews, don't be alarmed.
